File: exam_latest/exam_LRX/work4.cpp
//
// Created by liryi on 24-11-24.
//4.分割获取谷子的各叶片，即把每个叶片独立分割，并用不同颜色表示。
#include <opencv2/opencv.hpp>
#include <iostream>

int main() {
    // 读取图像
    cv::Mat image = cv::imread("..\\images\\image4.png");

    if (image.empty()) {
        std::cerr << "Image not found!" << std::endl;
        return -1;
    }

    // 转换为灰度图像
    cv::Mat gray;
    cv::cvtColor(image, gray, cv::COLOR_BGR2GRAY);

    // 高斯模糊去噪
    cv::Mat blurred;
    cv::GaussianBlur(gray, blurred, cv::Size(5, 5), 0);

    // Canny边缘检测
    cv::Mat edges;
    cv::Canny(blurred, edges, 50, 150);

    // 查找轮廓
    std::vector<std::vector<cv::Point>> contours;
    cv::findContours(edges, contours, cv::RETR_EXTERNAL, cv::CHAIN_APPROX_SIMPLE);

    // 创建白色背景图像
    cv::Mat output = cv::Mat::ones(image.size(), image.type()) * 0; // 黑色背景

    // 为每个叶片分配不同的颜色并填充
    for (size_t i = 0; i < contours.size(); ++i) {
        // 随机生成一个颜色
        cv::Scalar color(rand() % 256, rand() % 256, rand() % 256);
        // 填充每个轮廓区域
        cv::drawContours(output, contours, static_cast<int>(i), color, cv::FILLED);
    }

    // 拼接原图与分割结果图像进行对比
    cv::Mat combined;
    cv::hconcat(image, output, combined);  // 横向拼接原图与分割结果图像

    // 显示结果
    cv::imshow("Comparison: Original vs Segmentation", combined);

    cv::waitKey(0);
    return 0;
}


//逻辑：
//图像读取与灰度转换： 读取输入图像并转化为灰度图像，以便后续处理。
//降噪处理： 通过高斯模糊平滑图像，减少噪声对边缘检测的干扰。
//边缘检测： 使用 Canny 算法检测图像中的边缘，突出叶片轮廓。
//轮廓检测： 使用 cv::findContours 获取每片叶片的轮廓。
//颜色填充：
//为每片叶片的轮廓随机分配颜色。
//使用 cv::drawContours 填充轮廓区域，生成分割后的叶片图像。
//结果可视化： 横向拼接原始图像与分割结果图像，直观展示分割效果。
//结果分析：
//输出图像显示了每片叶片以不同颜色标注，分割效果清晰直观。
//若某些叶片未正确分割或颜色交错，可能与边缘检测阈值或轮廓提取参数有关，可适当调整。
//拼接的对比图有助于验证分割结果是否准确。

File: exam_latest/exam_TB/work4.py
import cv2
import matplotlib.pyplot as plt
import numpy as np

# 读取输入图像
image = cv2.imread('image/image4.png')

# 转换为灰度图，便于后续处理
gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)

# 使用高斯模糊去噪，减少细小噪声对边缘检测的干扰
blurred = cv2.GaussianBlur(gray, (5, 5), 0)

# 使用 Canny 算法检测边缘，提取叶片的轮廓
edges = cv2.Canny(blurred, 50, 150)

# 提取所有的外部轮廓
contours, _ = cv2.findContours(edges, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)

# 创建一个与原图大小相同的白色背景，用于绘制分割结果
output = np.ones_like(image) * 255  # 全白背景

# 遍历所有轮廓，为每个叶片区域填充随机颜色
for i, contour in enumerate(contours):
    # 过滤掉面积非常小的轮廓（避免噪声干扰）
    if cv2.contourArea(contour) < 1:
        continue
    # 为每个叶片生成随机颜色
    color = np.random.randint(0, 255, 3).tolist()  # 随机生成RGB颜色
    # 使用填充函数将轮廓内的区域填充为随机颜色
    cv2.fillPoly(output, [contour], color)

# 转换结果图像为 uint8 类型以便显示
output = output.astype(np.uint8)

# 显示原始图像和分割结果
plt.figure(figsize=(12, 6))

# 显示原始图像
plt.subplot(1, 2, 1)
plt.imshow(cv2.cvtColor(image, cv2.COLOR_BGR2RGB))  # 转换为 RGB 格式便于显示
plt.title('Original Image')
plt.axis('off')

# 显示分割后的图像
plt.subplot(1, 2, 2)
plt.imshow(cv2.cvtColor(output, cv2.COLOR_BGR2RGB))  # 转换为 RGB 格式便于显示
plt.title('Segmentation Image')
plt.axis('off')

# 显示图像窗口
plt.show()

# 代码逻辑分析
# 1. 图像读取与灰度转换：
#    读取输入图像，并将其转换为灰度图，便于后续处理。
# 2. 图像去噪：
#    使用高斯模糊（GaussianBlur）减少图像中的噪声，优化边缘检测效果。
# 3. 边缘检测：
#    使用 Canny 算法提取叶片边缘，生成二值边缘图像。
# 4. 轮廓提取：
#    使用 cv2.findContours 提取所有外部轮廓，确保能够获取每个独立叶片的边界。
# 5. 分割与填充：
#    为每个叶片区域分配随机颜色，并使用 cv2.fillPoly 填充轮廓区域。
# 6. 可视化结果：
#    显示原始图像和分割结果，直观对比分割效果。
